File: 095_Unique_Binary_Search_Trees_II/095_1.java
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Queue;
import java.util.ArrayDeque;
import java.util.LinkedList;
import java.util.Stack;


class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

class Solution {
    public List<TreeNode> generateTrees(int start, int end) {
        List<TreeNode> res = new LinkedList<>();
        if (start > end) {
            res.add(null);
            return res;
        }
        for (int i = start; i <= end; i++) {
            List<TreeNode> leftTrees  = generateTrees(start, i - 1);
            List<TreeNode> rightTrees = generateTrees(i + 1, end);
            for (TreeNode l : leftTrees) {
                for (TreeNode r : rightTrees) {
                    TreeNode s = new TreeNode(i);
                    s.left = l;
                    s.right = r;
                    res.add(s);
                }
            }
        }
        return res;
    }


    public List<TreeNode> generateTrees(int n) {
        if (n == 0)
            return new LinkedList<TreeNode>();
        else
            return generateTrees(1, n);
    }
}


public class leetcode {
	public static void main(String[] args) {
		Solution s =  new Solution();
		List<TreeNode> res = s.generateTrees(3);
		for (TreeNode r : res) {  // traversal trees
			Queue<TreeNode> queue = new ArrayDeque<>();
        	queue.add(r);
        	while (!queue.isEmpty()) {
        		int count = queue.size();
        		while (count-- > 0) {
        			TreeNode t = queue.poll();
        			System.out.print(t.val + " ");
        			if (t.left != null) queue.add(t.left);
        			if (t.right != null) queue.add(t.right);
        		}
        	}
        	System.out.println();
		}
	}
}

File: 095_Unique_Binary_Search_Trees_II/095_1.cpp
#include <vector>
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x): val(x), left(NULL), right(NULL) { }
};


class Solution {  // dynamic programming
public:
    vector<TreeNode*> generateTrees(int n) {
        vector<TreeNode *> res;
        if (n == 0)
            return res;
        res.push_back(nullptr);
        for (int i = 1; i <= n; i++) {
            vector<TreeNode *> cur;
            for (const auto & oldNode : res) {
                TreeNode *newNode = new TreeNode(i);
                newNode->left = oldNode;
                cur.push_back(newNode);
                for (int j = 0; j < i - 1; ++j) {
                    TreeNode *oldNodeCopy = treeCopy(oldNode);
                    TreeNode *rightNode = oldNodeCopy;
                    for (int k = 0; k < j; ++k) {
                        if (rightNode == nullptr) break;
                        rightNode = rightNode->right;
                    }
                    if (rightNode == nullptr) break;
                    TreeNode *rightTree = rightNode->right;
                    newNode = new TreeNode(i);
                    rightNode->right = newNode;
                    newNode->left = rightTree;
                    cur.push_back(oldNodeCopy);
                }
            }
            res = cur;
        }
        return res;
    }

    static TreeNode * treeCopy(TreeNode *oldNode) {
        if (oldNode == nullptr) return oldNode;
        TreeNode *newNode = new TreeNode(oldNode->val);
        newNode->left = treeCopy(oldNode->left);
        newNode->right = treeCopy(oldNode->right);
        return newNode;
    }

    static void traverse(TreeNode * tree) {
        cout << tree->val << " ";
        if (tree->left)  traverse(tree->left);
        if (tree->right) traverse(tree->right);
    }
};



int main() {
    Solution s;
    vector<TreeNode *> trees = s.generateTrees(3);
    for (const auto & t : trees) {
        Solution::traverse(t);
        cout << "\n";
    }
}